### ISSUE TITLE
[VoodooPS2] Keyboard not working (Trackpad is !)

### ISSUE THREAD

**Issue by jamesst20** [acidanthera/bugtracker#1640]
OS: Big Sur 11.3.1
Type: Laptop
Model: Alienware M17 R4 (2021)
Bootloader: OpenCore 0.6.9

What I have tried:

VoodooPS2Controller: Enables the touchpad (laggy) but no keyboard.
VoodoL2C: Enables the touchpad (butter smooth) but no keyboard.
VoodooPS2Controller (ReHabMan version): Still no keyboard

I am sorry if I failed to follow proper instructions. I tried to give as much information as I could

[Dump, ioreg and EFI.zip](https://github.com/acidanthera/bugtracker/files/6480115/Dump.and.EFI.zip)

[DSDT_SSDTTime_Windows.aml.zip](https://github.com/acidanthera/bugtracker/files/6480120/DSDT_SSDTTime_Windows.aml.zip)

Device Manager information (From Windows)
![Annotation 2021-05-13 233006](https://user-images.githubusercontent.com/490234/118302271-ff473580-b4b1-11eb-8ea1-cf4b75730b34.png)
![Annotation 2021-05-13 232628](https://user-images.githubusercontent.com/490234/118302272-ffdfcc00-b4b1-11eb-9dd4-e3c0c3a0eb04.png)
![Annotation 2021-05-13 232607](https://user-images.githubusercontent.com/490234/118302273-ffdfcc00-b4b1-11eb-8611-9e6042d6815d.png)
![Annotation 2021-05-13 232506](https://user-images.githubusercontent.com/490234/118302274-ffdfcc00-b4b1-11eb-850d-e109e25f9683.png)

**Comment by vit9696** [comment]
Sounds like your keyboard may be unsupported. Try https://github.com/coolstar/HIDPS2Keyboard. If nothing works, feel free to provide a patch.

**Comment by jamesst20** [comment]
> Sounds like your keyboard may be unsupported. Try https://github.com/coolstar/HIDPS2Keyboard. If nothing works, feel free to provide a patch.

Thank you for taking the time to reply to me. I built the suggested kext and tried it, unfortunately, no result on my end :/

I know C, I know C++, as much as I would love to contribute and provide a patch, I do not have any knowledge when it comes to writing driver. To be honest, I opened up VoodooPS2Keyboard.cpp and I am completely lost. 

Since I am such a specific case scenario,  I would understand you are not willing to invest time into it, however, I really love this laptop and if there is any motivation I can provide you for your time that I can afford, I would love to be able to type from this laptop this summer in the sun :) 

Let me know and thanks again. The hackintosh community wouldn't be what it is without you.

**Comment by vit9696** [comment]
Could you install DebugEnhancer and VoodooPS2Keyboard kexfs, and run sudo dmesg right after booting? Might help with the logging...

**Comment by jamesst20** [comment]
> Could you install DebugEnhancer and VoodooPS2Keyboard kexts, and run sudo dmesg right after booting? Might help with the logging...

Yes of course ! Anything you want or need. If this work (even if it don't) and you've got a donate button, I will be generous.

[dmesg.txt](https://github.com/acidanthera/bugtracker/files/6486703/dmesg.txt)

**Comment by vit9696** [comment]
Is that a DEBUG build?

**Comment by jamesst20** [comment]
> Is that a DEBUG build? Need to be one, sorry if it was not clear.
> 
> Press a few keys, especially caps lock a couple of times before dumping the log.

Sorry about that. It was the release version, it should have been obvious.

[dmesg2.txt](https://github.com/acidanthera/bugtracker/files/6487851/dmesg2.txt)

**Comment by vit9696** [comment]
```
[   49.836627]: ApplePS2Keyboard: sending key 1d=3b down
[   49.836631]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9a7dec1c, cmp=e0028010
[   49.848787]: VoodooPS2Keyboard: Unhandled keycode: 0x15B
[   49.848790]: ApplePS2Keyboard: sending key e05b=37 down
[   49.848794]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9b37c42b, cmp=e0028010
[   49.855238]: VoodooPS2Keyboard: Unhandled keycode: 0x76
[   49.855258]: ApplePS2Keyboard: sending key 76=80 down
[   49.855261]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9b99fe1c, cmp=e0028010
[   49.855283]: ApplePS2Keyboard::message: type=e3ff806e, provider=<ptr>, argument=<ptr>, argument=9b99fe1c, cmp=e0028010
[   49.862753]: VoodooPS2Keyboard: Unhandled keycode: 0x76
[   49.862756]: ApplePS2Keyboard: sending key 76=80 up
[   49.862760]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9c0cde9b, cmp=e0028010
[   49.862763]: ApplePS2Keyboard::message: type=e3ff806e, provider=<ptr>, argument=<ptr>, argument=9c0cde9b, cmp=e0028010
[   49.873685]: VoodooPS2Keyboard: Unhandled keycode: 0x15B
[   49.873689]: ApplePS2Keyboard: sending key e05b=37 up
[   49.873693]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9cb386c3, cmp=e0028010
[   49.878239]: VoodooPS2Keyboard: Unhandled keycode: 0x1D
[   49.878243]: ApplePS2Keyboard: sending key 1d=3b up
[   49.878245]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9cf90d67, cmp=e0028010
```

Your keyboard does provide input, and I do not see any issues with it. You can try seeing it online via `log stream` command, but it really looks just fine. Perhaps check that the scan codes match the keys you press, maybe your keyboard sends something unusual.

Otherwise it feels more like macOS itself is ignoring the keyboard, i.e. it is not the kext that causes the issue. Did you try keyboard input in macOS recovery?

CC @zhen-zen @usr-sse2

**Comment by zhen-zen** [comment]
Can you try https://karabiner-elements.pqrs.org and use `Karabiner-EventViewer.app` to check the input? I used to extract the app with https://mothersruin.com/software/SuspiciousPackage/ as there's no need to install the driver.

**Comment by jamesst20** [comment]
> ```
> [   49.836627]: ApplePS2Keyboard: sending key 1d=3b down
> [   49.836631]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9a7dec1c, cmp=e0028010
> [   49.848787]: VoodooPS2Keyboard: Unhandled keycode: 0x15B
> [   49.848790]: ApplePS2Keyboard: sending key e05b=37 down
> [   49.848794]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9b37c42b, cmp=e0028010
> [   49.855238]: VoodooPS2Keyboard: Unhandled keycode: 0x76
> [   49.855258]: ApplePS2Keyboard: sending key 76=80 down
> [   49.855261]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9b99fe1c, cmp=e0028010
> [   49.855283]: ApplePS2Keyboard::message: type=e3ff806e, provider=<ptr>, argument=<ptr>, argument=9b99fe1c, cmp=e0028010
> [   49.862753]: VoodooPS2Keyboard: Unhandled keycode: 0x76
> [   49.862756]: ApplePS2Keyboard: sending key 76=80 up
> [   49.862760]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9c0cde9b, cmp=e0028010
> [   49.862763]: ApplePS2Keyboard::message: type=e3ff806e, provider=<ptr>, argument=<ptr>, argument=9c0cde9b, cmp=e0028010
> [   49.873685]: VoodooPS2Keyboard: Unhandled keycode: 0x15B
> [   49.873689]: ApplePS2Keyboard: sending key e05b=37 up
> [   49.873693]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9cb386c3, cmp=e0028010
> [   49.878239]: VoodooPS2Keyboard: Unhandled keycode: 0x1D
> [   49.878243]: ApplePS2Keyboard: sending key 1d=3b up
> [   49.878245]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=9cf90d67, cmp=e0028010
> ```
> 
> 
>     
>   
> 
> Your keyboard does provide input, and I do not see any issues with it. You can try seeing it online via `log stream` command to confirm that it continues doing it during the normal functioning of the system, but it really looks just fine from the logs. Perhaps check that the scan codes match the keys you press, maybe your keyboard sends something unusual.
> 
> 
> Otherwise it feels more like macOS itself is ignoring the keyboard, i.e. it is not the kext that causes the issue. Did you try keyboard input in macOS recovery?
> 
> 
> CC @zhen-zen @usr-sse2

I did further testing. It seems the only way I can generate "ApplePS2Keyboard: sending key..." output is by pressing Fn+F11 or Fn+F1. The other keys don't generate any messages.

Also, Fn+F11 alone generate this whole output:

```
[ 1113.365498]: ApplePS2Keyboard: sending key 1d=3b down
[ 1113.365502]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=39c0b315, cmp=e0028010
[ 1113.380843]: VoodooPS2Keyboard: Unhandled keycode: 0x15B
[ 1113.380846]: ApplePS2Keyboard: sending key e05b=37 down
[ 1113.380849]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=3aab11cd, cmp=e0028010
[ 1113.385642]: VoodooPS2Keyboard: Unhandled keycode: 0x76
[ 1113.385644]: ApplePS2Keyboard: sending key 76=80 down
[ 1113.385659]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=3af452c0, cmp=e0028010
[ 1113.385662]: ApplePS2Keyboard::message: type=e3ff806e, provider=<ptr>, argument=<ptr>, argument=3af452c0, cmp=e0028010
[ 1113.391568]: VoodooPS2Keyboard: Unhandled keycode: 0x76
[ 1113.391570]: ApplePS2Keyboard: sending key 76=80 up
[ 1113.391572]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=3b4ebd7f, cmp=e0028010
[ 1113.391575]: ApplePS2Keyboard::message: type=e3ff806e, provider=<ptr>, argument=<ptr>, argument=3b4ebd7f, cmp=e0028010
[ 1113.403856]: VoodooPS2Keyboard: Unhandled keycode: 0x15B
[ 1113.403859]: ApplePS2Keyboard: sending key e05b=37 up
[ 1113.403861]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=3c0a3c99, cmp=e0028010
[ 1113.408507]: VoodooPS2Keyboard: Unhandled keycode: 0x1D
[ 1113.408509]: ApplePS2Keyboard: sending key 1d=3b up
[ 1113.408529]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=3c50ee8a, cmp=e0028010
```

and Fn+F1

```
[ 1268.768341]: VoodooPS2Keyboard: Unhandled keycode: 0x140
[ 1268.768346]: ApplePS2Keyboard: sending key e040=80 down
[ 1268.768350]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=687bff3b, cmp=e0028010
[ 1268.768353]: ApplePS2Keyboard::message: type=e3ff806e, provider=<ptr>, argument=<ptr>, argument=687bff3b, cmp=e0028010
[ 1268.778715]: VoodooPS2Keyboard: Unhandled keycode: 0x140
[ 1268.778717]: ApplePS2Keyboard: sending key e040=80 up
[ 1268.778718]: ApplePS2Keyboard::message: type=e3ff8066, provider=<ptr>, argument=<ptr>, argument=691a7268, cmp=e0028010
[ 1268.778721]: ApplePS2Keyboard::message: type=e3ff806e, provider=<ptr>, argument=<ptr>, argument=691a7268, cmp=e0028010
```



> Can you try https://karabiner-elements.pqrs.org and use `Karabiner-EventViewer.app` to check the input? I used to extract the app with https://mothersruin.com/software/SuspiciousPackage/ as there's no need to install the driver.

Trying in a minute. Reporting back soon

**Comment by zhen-zen** [comment]
So from what I understand, there's no output when keys other than Fn+F1 and Fn+F11 is pressed?

Can you try booting into Linux (LiveCD) and test which input is generating events?

Check the input list with `xinput --list` and test with `xinput test ID` for specific device. Things could be complicated when it comes to gaming keyboard.

**Comment by jamesst20** [comment]
> So from what I understand, there's no output when keys other than Fn+F1 and Fn+F11 is pressed?
> 
> Can you try booting into Linux (LiveCD) and test which input is generating events?
> 
> Check the input list with `xinput --list` and test with `xinput test ID` for specific device. Things could be complicated when it comes to gaming keyboard.

The kext logging can detect both Fn+1 and Fn+F11, however, Karabiner EventViewer only sees Fn+F1.

I went ahead, booted up Ubuntu 20.10 Live CD and screenshotted the results

![Keyboard](https://user-images.githubusercontent.com/490234/118394464-74427880-b634-11eb-9745-f6e1d4fe4300.png)

It looks like the ID generating events is 14

**Comment by zhen-zen** [comment]
The device name sounds like a USB one? iirc there's no such personalized name for PS2 one. Check `lsusb` whether the keyboard is in the list. And maybe you need to check if the USB map is working correctly: https://dortania.github.io/OpenCore-Post-Install/usb/

**Comment by andreszerocross** [comment]
> > So from what I understand, there's no output when keys other than Fn+F1 and Fn+F11 is pressed?
> > Can you try booting into Linux (LiveCD) and test which input is generating events?
> > Check the input list with `xinput --list` and test with `xinput test ID` for specific device. Things could be complicated when it comes to gaming keyboard.
> 
> The kext logging can detect both Fn+1 and Fn+F11, however, Karabiner EventViewer only sees Fn+F11.
> 
> I went ahead, booted up Ubuntu 20.10 Live CD and screenshotted the results
> 
> ![Keyboard](https://user-images.githubusercontent.com/490234/118394464-74427880-b634-11eb-9745-f6e1d4fe4300.png)
> 
> It looks like the ID generating events is 14


[config.plist.zip](https://github.com/acidanthera/bugtracker/files/6492698/config.plist.zip)

Use this config.pist,, reboot and check your keyboard.

**Comment by jamesst20** [comment]
@andreszerocross 

I tried it (and noticed you disabled the USB port limit quirk in the config.plist) but it did not make any difference. (Fn+F11 still working though)

> The device name sounds like a USB one? iirc there's no such personalized name for PS2 one. Check `lsusb` whether the keyboard is in the list. And maybe you need to check if the USB map is working correctly: https://dortania.github.io/OpenCore-Post-Install/usb/

Looks like you made a good assumption:

![xinput and usb](https://user-images.githubusercontent.com/490234/118478236-29d9fe00-b6ff-11eb-964a-5455c8748b7d.png)

USB Controller:
![usb controller](https://user-images.githubusercontent.com/490234/118478349-49712680-b6ff-11eb-9d41-08b9689a18bb.png)

If I turn it off with `echo "0000:00:14.0" | sudo tee /sys/bus/pci/drivers/xhci_hcd/unbind` the keyboard stops working but not the trackpad.

I must admit I'm very supprised. I will look into my USB port as fast as possible today and report back.

Thanks again!

**Comment by jamesst20** [comment]
![Capture d’écran, le 2021-05-17 à 11 44 45](https://user-images.githubusercontent.com/490234/118517762-80e2d180-b705-11eb-9b9c-b4fd52a8e722.png)

Just an update: It's like the keyboard doesn't exist from Mac OS point of view. I tested all my ports individually by hand to find all of them, but I can't seem to find which port is suppose to drive the keyboard and if I were to apply this usb mapping (by excluding all unused ports), I would obviously exclude the keyboard.

**Comment by andreszerocross** [comment]
Go to Windows without using OpenCore,,

![image](https://user-images.githubusercontent.com/9614101/118526793-cb4a6980-b76a-11eb-8278-bf4d23731da9.png)

Go to device manager,, Remove USB mouse or external USB keyboard. if you see HID Keyboard then PS2, just double click and click General. You will see address port of USB keyboard (if your internal keyboard is using XHC BUs).

Sorry for bad english

**Comment by andreszerocross** [comment]
Or you can use USBView.exe in windows to check address of your internal keyboard if it's using XHC Bus. It must be in USB input Device.

https://www.nirsoft.net/utils/usb_devices_view.html

**Comment by jamesst20** [comment]
> Or you can use USBView.exe in windows to check address of your internal keyboard if it's using XHC Bus. It must be in USB input Device.
> 
> https://www.nirsoft.net/utils/usb_devices_view.html

Thanks for the quick reply! It looks like it is not running on the XHC Bus. I attached the report from Nirsoft's USB Device View.


[Reports.zip](https://github.com/acidanthera/bugtracker/files/6495620/Reports.zip)
![USB](https://user-images.githubusercontent.com/490234/118529897-05d3e800-b712-11eb-9adb-51ce4beef5af.PNG)

**Comment by andreszerocross** [comment]
> > Or you can use USBView.exe in windows to check address of your internal keyboard if it's using XHC Bus. It must be in USB input Device.
> > https://www.nirsoft.net/utils/usb_devices_view.html
> 
> Thanks for the quick reply! It looks like it is not running on the XHC Bus. I attached the report from Nirsoft's USB Device View.
> 
> [Reports.zip](https://github.com/acidanthera/bugtracker/files/6495620/Reports.zip)
> ![USB](https://user-images.githubusercontent.com/490234/118529897-05d3e800-b712-11eb-9adb-51ce4beef5af.PNG)

![image](https://user-images.githubusercontent.com/9614101/118537984-cdff8b80-b777-11eb-8f9e-83f1b6ddc9ff.png)

This one is keyboard, right click and select View Device Detail..

**Comment by andreszerocross** [comment]
![image](https://user-images.githubusercontent.com/9614101/118539743-fb4d3900-b779-11eb-8fc2-85352bac69d3.png)

Sorry, i have check your reports.zip. I checked, and i can see your Keyboard is from HS10.

Just include this port (HS10) in your USB Map.. it looks your USB Map is exclude this port (HS10). Try to include this HS10 in your uSB Map and check your keyboard again.. it's better to upload IOREG output

**Comment by andreszerocross** [comment]
![](https://user-images.githubusercontent.com/490234/118519001-b3d99500-b706-11eb-848c-97ac91fddb99.png)

Your current USB Map is exclude HS10 port, sorry for bad english. Try to enable HS10

**Comment by jamesst20** [comment]
> ![image](https://user-images.githubusercontent.com/9614101/118539743-fb4d3900-b779-11eb-8fc2-85352bac69d3.png)
> 
> Sorry, i have check your reports.zip. I checked, and i can see your Keyboard is from HS10.
> 
> Just include this port (HS10) in your USB Map.. it looks your USB Map is exclude this port (HS10). Try to include this HS10 in your uSB Map and check your keyboard again.. it's better to upload IOREG output

OMG! Good Catch!!! I had downgraded to Big sur 11.2.3 to use XhciPortLimit patch. I also had included USBInjectAll. When I took those screenshots, I had no USB Map yet. HS10 was simply not detected. 

After some research, I hit this SSDT https://github.com/dortania/Getting-Started-With-ACPI/blob/master/extra-files/decompiled/SSDT-RHUB.dsl which is meant to force detect ports that are not present in the ACPI. Installed this and rebooted and boom! Keyboard is working! I will do proper USB patch now :)

Thank you so much to everyone !! It's so much appreciated and if I ever can help any of you in anyway. it will be my pleasure!! :) 

PS: Your English was perfectly fine :) and I'm also not a native English speaker (My native is French)
